pkg/instancetype/revision/store.go

```diff
@@ -0,0 +1,237 @@
+	vmiSpecCopy := vmiSpec.DeepCopy()
+	conflicts := apply.NewVMIApplier().ApplyToVMI(field.NewPath("spec", "template", "spec"), instancetypeSpec, nil, vmiSpecCopy, vmiMetadata)
+	if len(conflicts) > 0 {
+		return fmt.Errorf(instancetypeErrors.VMFieldsConflictsErrorFmt, conflicts.String())
+	}
+	return nil
+}
+
+func (h *RevisionHandler) storePreferenceRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+	if vm.Spec.Preference == nil || vm.Spec.Preference.RevisionName != "" {
+		return nil, nil
+	}
+
+	storedRevision, err := h.createPreferenceRevision(vm)
+	if err != nil {
+		return nil, err
+	}
+
+	vm.Spec.Preference.RevisionName = storedRevision.Name
+	return storedRevision, nil
+}
+
+func (h *RevisionHandler) createPreferenceRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+	switch strings.ToLower(vm.Spec.Preference.Kind) {
+	case api.SingularPreferenceResourceName, api.PluralPreferenceResourceName:
+		preference, err := preferenceFind.NewPreferenceFinder(h.preferenceStore, h.virtClient).Find(vm)
+		if err != nil {
+			return nil, err
+		}
+		return h.storeControllerRevision(vm, preference)
+	case api.ClusterSingularPreferenceResourceName, api.ClusterPluralPreferenceResourceName:
+		clusterPreference, err := preferenceFind.NewClusterPreferenceFinder(h.clusterPreferenceStore, h.virtClient).Find(vm)
+		if err != nil {
+			return nil, err
+		}
+		return h.storeControllerRevision(vm, clusterPreference)
+	default:
+		return nil, fmt.Errorf("got unexpected kind in PreferenceMatcher: %s", vm.Spec.Preference.Kind)
+	}
+}
+
+func GenerateName(vmName, resourceName, resourceVersion string, resourceUID types.UID, resourceGeneration int64) string {
+	return fmt.Sprintf("%s-%s-%s-%s-%d", vmName, resourceName, resourceVersion, resourceUID, resourceGeneration)
+}
+
+func CreateControllerRevision(vm *virtv1.VirtualMachine, object runtime.Object) (*appsv1.ControllerRevision, error) {
+	obj, err := util.GenerateKubeVirtGroupVersionKind(object)
+	if err != nil {
+		return nil, err
+	}
+	metaObj, ok := obj.(metav1.Object)
+	if !ok {
+		return nil, fmt.Errorf("unexpected object format returned from GenerateKubeVirtGroupVersionKind")
+	}
+
+	revisionName := GenerateName(
+		vm.Name, metaObj.GetName(),
+		obj.GetObjectKind().GroupVersionKind().Version,
+		metaObj.GetUID(),
+		metaObj.GetGeneration(),
+	)
+
+	// Removing unnecessary metadata
+	metaObj.SetLabels(nil)
+	metaObj.SetAnnotations(nil)
+	metaObj.SetFinalizers(nil)
+	metaObj.SetOwnerReferences(nil)
+	metaObj.SetManagedFields(nil)
+
+	return &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            revisionName,
+			Namespace:       vm.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(vm, virtv1.VirtualMachineGroupVersionKind)},
+			Labels: map[string]string{
+				api.ControllerRevisionObjectGenerationLabel: fmt.Sprintf("%d", metaObj.GetGeneration()),
```

Comment:
nit: Could be `strconv.Itoa(metaObj.GetGeneration())`

Reply:
https://github.com/kubevirt/kubevirt/issues/13044